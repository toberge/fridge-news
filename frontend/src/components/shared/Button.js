// @flow

import * as React from 'react';
import { Component } from 'react-simplified/lib/index';

class ButtonPrimary extends Component<{
  onClick: () => mixed,
  children?: React.Node,
  disabled?: boolean
}> {
  render() {
    return (
      <button type="button" className="btn btn-primary" onClick={this.props.onClick} disabled={this.props.disabled}>
        {this.props.children}
      </button>
    );
  }
}

class ButtonSuccess extends Component<{
  onClick: () => mixed,
  children?: React.Node
}> {
  render() {
    return (
      <button type="button" className="btn btn-success" onClick={this.props.onClick}>
        {this.props.children}
      </button>
    );
  }
}

class ButtonLight extends Component<{
  onClick: () => mixed,
  children?: React.Node
}> {
  render() {
    return (
      <button type="button" className="btn btn-light" onClick={this.props.onClick}>
        {this.props.children}
      </button>
    );
  }
}

class ButtonDanger extends Component<{
  onClick: () => mixed,
  children?: React.Node
}> {
  render() {
    return (
      <button type="button" className="btn btn-danger" onClick={this.props.onClick}>
        {this.props.children}
      </button>
    );
  }
}

/**
 * Button widgets
 */
export default class Button {
  static Success = ButtonSuccess;
  static Secondary = ButtonLight;
  static Danger = ButtonDanger;
  static Primary = ButtonPrimary;
}
