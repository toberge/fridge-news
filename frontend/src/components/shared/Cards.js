// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { NavLink } from 'react-router-dom';
import { ArticleBase, capitalizeFirstLetter } from '../../data/Article';
import placeholderImage from '../../assets/images/floppy.jpg';
// Current placeholder is public domain and does not require contribution.
// Regardless, here's where I found it: https://www.pexels.com/photo/office-disk-storage-data-41290/
// and where it seems to be from: https://www.publicdomainpictures.net/en/view-image.php?image=1454

export class CardHolder extends Component<{ children: React.Node }> {
  render() {
    return <div className="card-columns">{this.props.children}</div>;
  }
}

export class ArticleCard extends Component<{
  article: ArticleBase,
  showCategory?: boolean
}> {
  render() {
    return (
      <NavLink className="card" style={{ width: '40rem;' }} to={`/articles/${this.props.article.id}`}>
        <img
          className="card-img-top"
          src={this.props.article.picturePath ? this.props.article.picturePath : placeholderImage}
          alt={this.props.article.pictureAlt ? this.props.article.pictureAlt : 'Floppy disks'}
        />
        <div className="card-body">
          <h2 className="card-title">{this.props.article.title}</h2>
        </div>
        {this.props.showCategory ? (
          <div className="card-footer text-muted">{capitalizeFirstLetter(this.props.article.category)}</div>
        ) : null}
      </NavLink>
    );
  }
}

export class NewsFeedCard extends Component<{
  article: ArticleBase
}> {
  render() {
    return (
      <div className="card">
        <div className="card-body">
          <div className="card-text">
            <strong>
              <NavLink className="card" to={`/articles/${this.props.article.id}`}>
                {this.props.article.title}
              </NavLink>
            </strong>
          </div>
        </div>
      </div>
    );
  }
}
