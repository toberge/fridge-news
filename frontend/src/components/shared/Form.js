// @flow

import * as React from 'react';
import { Component } from 'react-simplified';

class FormSubmit extends Component<{
  disabled: boolean,
  children?: React.Node
}> {
  render() {
    return (
      <button type="submit" className="btn btn-primary" disabled={this.props.disabled}>
        {this.props.children}
      </button>
    );
  }
}

class FormInput extends Component<{
  type?: string,
  name: string,
  label: string,
  placeholder: string,
  value: string,
  onChange: (SyntheticInputEvent<HTMLInputElement>) => void,
  required?: boolean
}> {
  render() {
    return (
      <div className="form-group">
        <label htmlFor={this.props.name} className="form-label">
          {this.props.label}
        </label>
        <input
          id={this.props.name}
          className="form-control"
          placeholder={this.props.placeholder}
          type={this.props.type ? this.props.type : 'text'}
          value={this.props.value}
          onChange={this.props.onChange}
          required={this.props.required}
        />
      </div>
    );
  }
}

class FormInputRow extends Component<{
  type?: string,
  name: string,
  label: string,
  placeholder: string,
  helpText: string | React.Node,
  value: string,
  onChange: (SyntheticInputEvent<HTMLInputElement>) => void,
  required?: boolean,
  labelCols?: number,
  inputCols?: number
}> {
  render() {
    return (
      <div className="row">
        <label htmlFor={this.props.name} className={`col${this.props.labelCols ? '-' + this.props.labelCols : ''} col-form-label`}>
          {this.props.label}
        </label>
        <div className={`col-${this.props.inputCols ? this.props.inputCols : 10}`}>
          <input
            id={this.props.name}
            aria-describedby={`${this.props.name}Help`}
            className="form-control"
            placeholder={this.props.placeholder}
            type={this.props.type ? this.props.type : 'text'}
            value={this.props.value}
            onChange={this.props.onChange}
            required={this.props.required}
          />
          <small id={`${this.props.name}Help`} className="form-text text-muted">
            {this.props.helpText}
          </small>
        </div>
      </div>
    );
  }
}

class FormGroup extends Component<{ children: React.Node }> {
  render() {
    return <div className="form-group">{this.props.children}</div>;
  }
}

export default class Form extends Component<{ children: React.Node }> {
  static Input = FormInputRow;
  static Group = FormGroup;
  static Submit = FormSubmit;

  render() {
    return <form>{this.props.children}</form>;
  }
}
