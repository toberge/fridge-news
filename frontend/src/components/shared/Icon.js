// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import './Icon.css';

class LoadingIcon extends Component {
  render() {
    return <i className="fas fa-spinner fa-fw fa-spin"></i>;
  }
}

class WriteIcon extends Component {
  render() {
    return <i className="fas fa-edit"></i>;
  }
}

class SaveIcon extends Component {
  render() {
    return <i className="fas fa-save"></i>;
  }
}

class UploadIcon extends Component {
  render() {
    return <i className="fas fa-upload"></i>;
  }
}

class SearchIcon extends Component {
  render() {
    return <i className="fas fa-search"></i>;
  }
}

class UserIcon extends Component {
  render() {
    return <i className="fas fa-user"></i>;
  }
}

class EmailIcon extends Component {
  render() {
    return <i className="far fa-envelope"></i>;
  }
}

class DeleteIcon extends Component {
  render() {
    return <i className="fas fa-trash"></i>;
  }
}

export default class Icon {
  static Loading = LoadingIcon;
  static Write = WriteIcon;
  static Upload = UploadIcon;
  static Save = SaveIcon;
  static Delete = DeleteIcon;
  static Search = SearchIcon;
  static User = UserIcon;
  static Email = EmailIcon;
}
