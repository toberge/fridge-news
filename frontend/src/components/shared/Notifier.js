// @flow

import * as React from 'react';
import notifier from 'simple-react-notifications';
import "simple-react-notifications/dist/index.css";
import "./Notifier.css";

notifier.configure({
  position: 'top-center',
  closeOnClick: true,
  // width: '30rem', TODO needs further styling... X shows up too far left
  animation: {
    in: "fadeIn",
    out: "fadeOut",
    duration: 200
  }
});

export default class Notifier {
  static notifs: number[] = [];

  static error(msg: string) {
    this.notifs.push(notifier.error(msg, {
      autoClose: 4000
    }));
  }

  static info(msg: string) {
    this.notifs.push(notifier.info(msg));
  }

  static success(msg: string) {
    this.notifs.push(notifier.success(msg));
  }
}
