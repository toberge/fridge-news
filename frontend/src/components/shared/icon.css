.fa-spin-slow {
  -webkit-animation: fa-spin 4s infinite ease;
  -o-animation: fa-spin 4s infinite ease;
  animation: fa-spin 4s infinite ease;
}
