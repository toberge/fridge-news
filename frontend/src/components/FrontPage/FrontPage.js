// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { ArticleCard, CardHolder } from '../shared/Cards';
import { articleStore } from '../../stores/articleStore';

export default class FrontPage extends Component {
  render() {
    return (
      <main>
        <h1>Fridge News</h1>
        <CardHolder>
          {articleStore.articles.map(a => (
            <ArticleCard article={a} key={a.id} showCategory />
          ))}
        </CardHolder>
      </main>
    );
  }

  mounted(): void {
    articleStore.getFrontPage().catch(error => console.error(error));
    document.title = 'Fridge News | Home';
  }
}
