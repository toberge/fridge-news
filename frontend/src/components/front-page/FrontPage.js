// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { ArticleCard, CardHolder } from '../shared/Cards';
import { articleStore } from '../../stores/articleStore';
import Notifier from '../shared/Notifier';

export default class FrontPage extends Component {
  render() {
    return (
      <main>
        <h1>Fridge News</h1>
        <CardHolder>
          {articleStore.articles.map(a => (
            <ArticleCard article={a} key={a.id} showCategory />
          ))}
        </CardHolder>
      </main>
    );
  }

  mounted(): void {
    articleStore.getFrontPage().catch((error: Error) => Notifier.error(`Could not fetch articles\n${error.message}`));
    document.title = 'Fridge News';
  }
}
