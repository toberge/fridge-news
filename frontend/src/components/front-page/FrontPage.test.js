// @flow

import * as React from 'react';
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import { shallow, ShallowWrapper } from 'enzyme/build';
import { articleStore } from '../../stores/articleStore';
import { ArticleBase } from '../../data/Article';
import FrontPage from './FrontPage';
import { ArticleCard } from '../shared/Cards';

jest.mock('../../stores/articleStore');
// mock this promise
const storeSpy = jest.spyOn(articleStore, 'getFrontPage').mockResolvedValue();

describe('ArticleViewer', () => {
  // set up data
  articleStore.articles = [
    new ArticleBase(1, 'whatever', '', '', 'science'),
    new ArticleBase(2, 'some stuff', 'https://totallyanimage.com/p.jpg', 'alt alt', 'culture')
  ];

  const frontPage: ShallowWrapper = shallow(<FrontPage />);

  it('should render article list as intended', () => {
    expect(frontPage.debug()).toMatchSnapshot();
    expect(frontPage.find(ArticleCard).first().props().article.title).toBe('whatever');
  });

  it('should make articleStore fetch articles', () => {
    expect(storeSpy).toHaveBeenCalled();
  })
});
