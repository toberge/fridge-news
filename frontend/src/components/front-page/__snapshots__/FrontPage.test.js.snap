// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleViewer should render as intended 1`] = `
"<main>
  <h1>
    Fridge News
  </h1>
  <CardHolder>
    <ArticleCard article={{...}} showCategory={true} />
    <ArticleCard article={{...}} showCategory={true} />
  </CardHolder>
</main>"
`;
